servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpTransportObserverAsyncContextTest.java

```diff
@@ -0,0 +1,328 @@
+
+    public HttpTransportObserverAsyncContextTest(HttpProtocol protocol) {
+        super(CACHED, CACHED_SERVER);
+        this.protocol = protocol;
+        protocol(protocol.config);
+        transportObserver(clientObserver, NoopTransportObserver.INSTANCE);
+    }
+
+    @Parameters(name = "protocol={0}")
+    public static HttpProtocol[] data() {
+        return HttpProtocol.values();
+    }
+
+    @Before
+    public void setupContext() {
+        AsyncContext.put(CLIENT_KEY, CLIENT_VALUE);
+    }
+
+    @Test
+    public void echoRequestResponse() throws Exception {
+        String requestContent = "request_content";
+        testRequestResponse(newRequest(SVC_ECHO), OK, requestContent.length());
+    }
+
+    @Test
+    public void clientErrorWrite() {
+        StreamingHttpClient client = streamingHttpClient();
+        assertThrows(ExecutionException.class, () -> client.request(client.post(SVC_NO_CONTENT_AFTER_READ)
+                .payloadBody(Publisher.failed(DELIBERATE_EXCEPTION)))
+                .toFuture().get());
+        assertMap(clientObserver.storageMap(), CLIENT_VALUE);
+    }
+
+    @Test
+    public void serverHandlerError() throws Exception {
+        testRequestResponse(newRequest(SVC_THROW_ERROR), INTERNAL_SERVER_ERROR, 0);
+    }
+
+    @Test
+    public void serverErrorBeforeRead() throws Exception {
+        testServerReadError(SVC_ERROR_BEFORE_READ);
+    }
+
+    @Test
+    public void serverErrorDuringRead() throws Exception {
+        testServerReadError(SVC_ERROR_DURING_READ);
+    }
+
+    private StreamingHttpRequest newRequest(String path) {
+        String requestContent = "request_content";
+        return streamingHttpClient().post(path)
+                .addHeader(CONTENT_LENGTH, valueOf(requestContent.length()))
+                .payloadBody(getChunkPublisherFromStrings(requestContent));
+    }
+
+    private void testRequestResponse(StreamingHttpRequest request, HttpResponseStatus expectedStatus,
+                                     int expectedResponseLength) throws Exception {
+        StreamingHttpResponse response = streamingHttpClient().request(request).toFuture().get();
+        assertResponse(response, protocol.version, expectedStatus, expectedResponseLength);
+        assertMap(clientObserver.storageMap(), CLIENT_VALUE);
+    }
+
+    private void testServerReadError(String path) throws Exception {
+        StreamingHttpResponse response = streamingHttpClient().request(newRequest(path))
+                .toFuture().get();
+        assertResponse(response, protocol.version, OK);
+        assertThrows(ExecutionException.class, () -> response.payloadBody().ignoreElements().toFuture().get());
+        assertMap(clientObserver.storageMap(), CLIENT_VALUE);
+    }
+
+    private static void assertMap(Map<String, String> map, String expected) {
+        Set<Map.Entry<String, String>> entries = map.entrySet();
+        assertThat("No entries in the map", entries, hasSize(greaterThan(0)));
```

Comment:
nit `hasSize(greaterThan(0))` -> `not(Matchers.empty())` (prefer `isEmpty()` instead of `size()` if you don't need the actual size)